lib/Model/Importer/DebugInfo/PDBImporter.cpp

```diff
@@ -0,0 +1,953 @@
+//
+
+#include <csignal>
+#include <optional>
+
+#include "llvm/DebugInfo/CodeView/CVSymbolVisitor.h"
+#include "llvm/DebugInfo/CodeView/SymbolDeserializer.h"
+#include "llvm/DebugInfo/CodeView/SymbolVisitorCallbackPipeline.h"
+#include "llvm/DebugInfo/CodeView/TypeDumpVisitor.h"
+#include "llvm/DebugInfo/PDB/Native/DbiStream.h"
+#include "llvm/DebugInfo/PDB/Native/GlobalsStream.h"
+#include "llvm/DebugInfo/PDB/Native/ModuleDebugStream.h"
+#include "llvm/DebugInfo/PDB/Native/NativeSession.h"
+#include "llvm/DebugInfo/PDB/Native/PDBFile.h"
+#include "llvm/DebugInfo/PDB/Native/SymbolStream.h"
+
+#include "revng/Model/Importer/DebugInfo/PDBImporter.h"
+#include "revng/Model/Pass/AllPasses.h"
+#include "revng/Model/Processing.h"
+#include "revng/Model/QualifiedType.h"
+#include "revng/Model/Type.h"
+#include "revng/Support/Assert.h"
+#include "revng/Support/Debug.h"
+#include "revng/Support/MetaAddress.h"
+
+using namespace llvm;
+
+static Logger<> DILogger("pdb-importer");
+
+#include <iostream>
+
+void PDBImporter::import(const object::COFFObjectFile &TheBinary) {
+  using namespace llvm::object;
+
+  // Parse debug info and populate types to Model.
+  const codeview::DebugInfo *DebugInfo;
+  StringRef PDBFilePath;
+  using namespace llvm::pdb;
+  static std::unique_ptr<IPDBSession> Session;
+  auto EC = TheBinary.getDebugPDBInfo(DebugInfo, PDBFilePath);
+  if (!EC && DebugInfo != nullptr && !PDBFilePath.empty()) {
+    uint64_t PDBSignature = DebugInfo->Signature.CVSignature;
+
+    if (llvm::sys::fs::exists(PDBFilePath)) {
+      auto Err = loadDataForPDB(PDB_ReaderType::Native, PDBFilePath, Session);
+      if (!Err)
+        ThePDBFile = &static_cast<NativeSession *>(Session.get())->getPDBFile();
+      else
+        revng_log(DILogger, "Unable to read PDB file.");
+    } else {
+      // Usualy the PDB files will be generated on a different machine,
+      // so the location wont be up to date. Try to find it:
+      //  1. in the current `.` dir
+      //  2. in an user defined/provided location???
+      //  3. somewhere else ???
+      auto PDBFileNameOnly = PDBFilePath.substr(PDBFilePath.find_last_of('\\')
+                                                + 1);
+      if (llvm::sys::fs::exists(PDBFileNameOnly.str())) {
+        auto Err = loadDataForPDB(PDB_ReaderType::Native,
+                                  PDBFileNameOnly.str(),
+                                  Session);
+        if (!Err)
+          ThePDBFile = &static_cast<NativeSession *>(Session.get())
+                          ->getPDBFile();
+        else
+          revng_log(DILogger, "Unable to read PDB file.");
+      } else {
+        revng_log(DILogger, "Unable to find PDB file.");
+      }
+
+      // TODO: Check the signature from PDB Info Stream. It should match
+      // with the PDBSignature from exe.
+    }
+  }
+
+  if (!ThePDBFile)
+    return;
+
+  populateTypes();
+  populateSymbolsWithTypes(*static_cast<NativeSession *>(Session.get()));
```

Comment:
* This should be methods of a local class `PDBImporterImpl` in order to avoid leaking implementation details into the public `.h` file.
* `static_cast`s are sad, let's have them in a single place with a comment why we need it.

Reply:
I agree. Thanks.